src/uct/tcp/tcp_cm.c

```diff
@@ -75,200 +82,426 @@ static ucs_status_t uct_tcp_cm_io_err_handler_cb(void *arg, int io_errno)
+uct_tcp_cm_simult_conn_accept_remote_conn(uct_tcp_ep_t *accept_ep,
+                                          uct_tcp_ep_t *connect_ep,
+                                          unsigned *progress_count)
 {
     uct_tcp_cm_conn_event_t event;
     ucs_status_t status;
 
-    status = ucs_socket_recv(ep->fd, &event, sizeof(event),
-                             uct_tcp_cm_io_err_handler_cb, ep);
+    /* 1. Close the allocated socket `fd` to avoid reading any
+     *    events for this socket and assign the socket `fd` returned
+     *    from `accept()` to the found EP */
+    uct_tcp_ep_mod_events(connect_ep, 0, connect_ep->events);
+    ucs_assertv(connect_ep->events == 0,
+                "Requsted epoll events must be 0-ed for ep=%p", connect_ep);
+
+    close(connect_ep->fd);
+    connect_ep->fd = accept_ep->fd;
+
+    /* 2. Migrate RX from the EP allocated during accepting connection to
+     *    the found EP */
+    status = uct_tcp_ep_move_ctx_cap(accept_ep, connect_ep,
+                                     UCT_TCP_EP_CTX_TYPE_RX);
     if (status != UCS_OK) {
-        uct_tcp_cm_trace_conn_pkt(ep, "unable to receive connection ack from",
-                                  &ep->peer_addr);
         return status;
     }
 
-    status = uct_tcp_cm_conn_pkt_check_event(ep, UCT_TCP_CM_CONN_ACK, event);
+    /* 3. Destroy the EP allocated during accepting connection
+     *    (set its socket `fd` to -1 prior to avoid closing this socket) */
+    uct_tcp_ep_mod_events(accept_ep, 0, EPOLLIN);
+    accept_ep->fd = -1;
+    uct_tcp_ep_destroy_internal(&accept_ep->super.super);
+    accept_ep = NULL;
+
+    /* 4. Send ACK to the peer */
+    event = UCT_TCP_CM_CONN_ACK;
+
+    /* 5. Tie REQ with ACK and send it to the peer using new socket fd to
+     *    ensure that the peer will be able to receive the data from us */
+    if (connect_ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING) {
+        event |= UCT_TCP_CM_CONN_REQ;
+    }
+
+    status = uct_tcp_cm_send_event(connect_ep, event);
     if (status != UCS_OK) {
         return status;
     }
 
-    uct_tcp_cm_trace_conn_pkt(ep, "connection ack received from",
-                              &ep->peer_addr);
+    (*progress_count)++;
+
+    /* 6. Ok, now fully connected to the peer */
+    uct_tcp_cm_change_conn_state(connect_ep, UCT_TCP_EP_CONN_STATE_CONNECTED);
+    uct_tcp_ep_mod_events(connect_ep, EPOLLIN | EPOLLOUT, 0);
+
     return UCS_OK;
 }
 
-unsigned uct_tcp_cm_conn_progress(uct_tcp_ep_t *ep)
+static ucs_status_t uct_tcp_cm_handle_simult_conn(uct_tcp_iface_t *iface,
+                                                  uct_tcp_ep_t *accept_ep,
+                                                  uct_tcp_ep_t *connect_ep,
+                                                  unsigned *progress_count)
 {
     ucs_status_t status;
+    int cmp;
 
-    status = ucs_socket_connect_nb_get_status(ep->fd);
+    cmp = ucs_sockaddr_cmp((const struct sockaddr*)&connect_ep->peer_addr,
```

Comment:
minor: you could compare addresses when not `UCT_TCP_EP_CONN_STATE_CONNECTED` only